### DISCUSSION TITLE
C Bindings (facilitating bindings from other static languages)

### DISCUSSION THREAD

**Discussion by geoffder** [elalish/manifold#284] [Ideas]
Out of a desire to be able to write OCaml bindings that will allow me to use Manifold as an alternative more "direct" backend for my mesh generation library [OCADml](https://github.com/OCADml/OCADml) (currently using OpenSCAD via [OSCADml](https://github.com/OCADml/OCADml)), I've begun work on some [Manifold C bindings](https://github.com/geoffder/manifold/tree/cbindings/bindings/c).

I've previously written [OCaml bindings](https://github.com/geoffder/olm-ml) to the [olm cryptographic library](https://gitlab.matrix.org/matrix-org/olm), which exposes a C API designed to facilitate binding from higher level languages, so I have tried to follow a similar pattern. Namely passing in memory allocated in the host language such that it can be managed appropriately.

I'm happy to open a PR as well, but I thought that I would link it here for some early feedback. It's been a while since I have written any C/C++!

#### Note:
I am using the export module, so the build should be run with:
```bash
cmake -DCMAKE_BUILD_TYPE=Release -DBUILD_SHARED_LIBS=ON -DMANIFOLD_EXPORT=ON .. && make
```
Also, I was getting an error due to deprecation of `pbrmaterial.h` so I made a couple small substitutions in `meshIO.cpp` to workaround it.

**Comment by elalish** [comment]
Thank you! I know basically nothing about C, so would you mind answering a few noob questions? What high level languages are you looking to target? We already have bindings for JS and Python, and found no need for a C API. How does the C API make it easier? I do want this library to be as easily usable as possible, so a PR would be great; my main concern is that without the C knowledge I'm concerned I won't be able to test/maintain it well. Of course, open source is all about collaboration, so if you're interested in being involved longer-term that would make a huge difference. 

Also curious what @pca006132 thinks? 

Regarding Assimp - we're using the old version only because that's what's available on ubuntu 20.04 on our CI - we could probably just upgrade to 22 to solve this.

**Reply by geoffder** [reply]
Great, glad to hear you're interested. As mentioned, the idea with C bindings is that they serve as a middle-man for bindings written in other languages (for me, that's OCaml). Oweing to its stable and much simpler interface and the fact that many high level static language runtimes are written in C, the FFI story is a lot better with C than C++. I'm not so familiar with the tooling that python and JS have available for binding C++, but I imagine the situation is a bit simpler for them being dynamic (on top of the enormous amount of man power they have that enables such tooling).

With how promising Manifold is, I'm sure it wouldn't be long until others are inspired to make bindings in their preferred languages if a C API is provided (I'll certainly be hosting some under OCADml). 

I'd be happy to stick around to help the C bindings keep up with the inevitable changes as the library continues to evolve.

**Reply by pca006132** [reply]
I am open to this, but I will need some time to review it later (I got covid after the paper deadline :P, how fortunate this is lol)

**Reply by elalish** [reply]
Oh no! Feel better soon @pca006132 - my whole family got it this summer. Definitely sucked, but no worse than a nasty flu for us at least. Hopefully better for you.

**Reply by elalish** [reply]
Thanks @geoffder, more hands are always appreciated! Is it just me or is everyone inventing a language these days? I hadn't even heard of OCaml, but it looks interesting.

**Reply by geoffder** [reply]
It's not so new (1996, a descendant of a line of ML family languages before that), but it certainly isn't all that well known 😅. It's probably going to take me a bit, but I think I'll get started on the OCaml bindings using my fork so I can discover any pain points with the C API and iterate.

**Reply by pca006132** [reply]
OCaml is a functional programming language, so it is probably less well-known to people not working on programming languages, but it is definitely not something new :). I think the C binding looks pretty clean, which should be good. Using this API directly (without compiler inserting delete or GC calling the destructor) might be a bit painful though, we can probably add something like a memory arena (https://github.com/elalish/manifold/blob/master/bindings/wasm/examples/worker.js#L67) that helps the user manage memory references in a less painful way.

**Reply by geoffder** [reply]
Thanks. It is true that it would be painful to use directly, but the intended purpose of the API is to make it straightforward to bind into from other languages (allocating the memory in the host language and registering a finalization for the GC). I'm not as familiar with them, but I did spend time looking at the olm bindings in Rust as I was writing mine in OCaml and it seemed very straightforward for them as well (more automated than mine in fact). That there is nothing fancy happening with the memory in the C API is actually a benefit here.

Fortunately I don't think there are so many interested in using a library like this through C directly.

**Reply by pca006132** [reply]
True, I guess we can merge the current C API later when you tried it with OCaml.

**Comment by geoffder** [comment]
So I've got a project going on the OCaml side that I hope I'll be able to push along more soon. While that's coming along I should ask about whether something could be done about the biggest rough edge of the C bindings (as marked by a comment around the decompose functions). In order to know how many manifolds worth of memory must be allocated for the result of decompose, the entire method has to be run. 

Would you be open to breaking out a method that ends with providing the indices/number of independent meshes, and have another method that takes those as input so they don't need to be recalculated? I didn't end up doing anything myself yet since I wasn't sure of the cleanest way to go about it that wouldn't dirty up the API, but also make the methods readily available to the C bindings.

**Reply by elalish** [reply]
Hmm, I may not have thought this through completely, but decompose is exactly what it says: it doesn't create or destroy anything, it just sorts out the pieces. All of the decomposed manifolds together should have exactly the same faces/verts (memory) as the single input manifold, since it's just a partition. Would that help? The expensive part is running the graph search, and that solves both the number of components and the mapping, so I don't think there's a way to separate the computations.

**Reply by geoffder** [reply]
To make sure that I'm following:
```cpp
  std::vector<int> components;
  const int numLabel = ConnectedComponents(components, graph);
```
This is the expensive part right? If I'm correct, that's where I was suggesting to break the method apart. With the knowledge of how big of a buffer to allocate (how many manifolds to be returned), then using the C binding can avoid doing the `ConnectedComponents` operation twice (returning the components vector from the first function such that it can be fed back in for the cheap part). For the C++ API of course it shouldn't be broken into the cumbersome two steps, so `Decompose` would bundle these together.

**Reply by elalish** [reply]
Yes, that's right. Okay, that seems reasonable. This reminds me how glad I am I never had to write C.

**Reply by geoffder** [reply]
Yea, I definitely don't prefer it either :sweat_smile:

**Comment by geoffder** [comment]
Another separate issue, I just tried to build with CUDA on for the first time, and I'm running into an issue building the C binding library (am able to build manifold itself with the python bindings fine). I think I've tracked it down to LevelSet and its function parameter that requires template arguments to work with CUDA.

I tried replacing this:
```cpp
ManifoldMesh *manifold_level_set(void *mem, float (*sdf)(ManifoldVec3),
                                 ManifoldBox *bounds, float edge_length,
                                 float level) {
  auto fun = [sdf](glm::vec3 v) { return (sdf(to_struct(v))); };
  auto mesh = LevelSet(fun, *from_c(bounds), edge_length, level);
  return to_c(new (mem) Mesh(mesh));
}

```
with this abomination to try and fool it:
```cpp
ManifoldMesh *manifold_level_set(void *mem, float (*sdf)(ManifoldVec3),
                                 ManifoldBox *bounds, float edge_length,
                                 float level) {
  struct F {
    float (*func)(ManifoldVec3);
    __host__ __device__ float operator()(glm::vec3 v) const {
      return (func(to_struct(v)));
    }
    F(float (*f)(ManifoldVec3)) : func(f){};
  };
  auto mesh = LevelSet(F(sdf), *from_c(bounds), edge_length, level);
  return to_c(new (mem) Mesh(mesh));
}
```
But of course it doesn't work, due to I guess what should have been obvious to me with the `const` keyword. I'm not so familiar with GPU stuff, and whether this is something that cannot be worked around dynamically (giving a function that was not defined completely statically in C++, but written using a passed in function pointer, the template type that it needs).

**Reply by elalish** [reply]
I dunno; I recall I tried to write the LevelSet API with a std::function and it didn't work. The basic idea is that the GPU requires its code to be compiled separately and the functions stored in a different memory space, so there's no way to pass them (they must be recompiled). Templating works because it ensures all of that happens at compile time - it's not possible at run time. Basically everything on the GPU is passed in by value, and I don't think there's such thing for functions in C.

**Reply by pca006132** [reply]
Yeah this problem is also present in the python binding. If I recall correctly, I just expose a version that doesn't perform GPU acceleration to users, but it can still be multithreaded. One possible way would be to JIT the function but it is too much work and idk if it is possible to do this nicely without too much code.

**Reply by geoffder** [reply]
Ah ok, will just have to add a rule that checks for the `CBIND` flag, and avoid trying to compile with the GPU templateing for `LevelSet` so that turning on GPU for the rest doesn't bring it all down. I haven't gone looking yet, but since the python bindings are dynamic I don't think they had to dodge the issue in the same way did they?

**Reply by geoffder** [reply]
I've just tested out the pattern of typing the function argument to `LevelSet` as a `std::function` rather than an `auto` lambda (similar to how the WASM binding passes a function pointer I think). While it does not *fail* compilation (as it does with a lambda as it was when mentioned previously), there is a warning, and it fails at runtime:
```
terminate called after throwing an instance of 'thrust::system::system_error'
  what():  parallel_for: failed to synchronize: cudaErrorLaunchFailure: unspecified launch failure
zsh: IOT instruction (core dumped)  ./bindings/c/testc
```
Unless I'm missing it, I don't see in the other bindings source or build files that attempts to avoid this. My understanding of the template stuff going on to use CUDA isn't too great, but I have the feeling a fix would have to go into `sdf.h` itself to change the types when bindings are being compiled (since none of them can satisfy the static function requirement it has (in its current form). Does that make sense to you guys?

**Reply by geoffder** [reply]
I have cooked up this hack for `LevelSet`:
```cpp
#ifdef AVOID_LAMBDA_CUDA
  const auto auto_policy = autoPolicy(maxMorton);
  const auto policy = auto_policy == ParUnseq ? Par : auto_policy;
#else
  const auto policy = autoPolicy(maxMorton);
#endif
```
with this in the top-level `CMakeLists.txt`:
```cmake
if(CBIND)
    # for avoiding handing lambda to for_each_n with a GPU policy in sdf LevelSet
    add_compile_definitions(AVOID_LAMBDA_CUDA)
endif()
```
I haven't tested them, but from my experience on the C side, I suspect this should probably be turned on for the other bindings as well.

**Reply by pca006132** [reply]
I think maybe we should just use template specialization: specialize the function parameter as std::function, and avoid using the GPU in that specialization.

**Reply by geoffder** [reply]
Ah good point, a specialized template wrapper around the call to `for_each_n` that avoids `ParUnseq` you think?

**Reply by pca006132** [reply]
Yes

**Reply by geoffder** [reply]
This works:
```cpp
template <typename Func>
void for_each_n_wrapper(ExecutionPolicy policy, int morton,
                        ComputeVerts<Func> cv) {
  for_each_n(policy, countAt(0), morton, cv);
}
template <>
void for_each_n_wrapper<std::function<float(glm::vec3)>>(
    ExecutionPolicy policy, int morton,
    ComputeVerts<std::function<float(glm::vec3)>> cv) {
  const auto pol = policy == ParUnseq ? Par : policy;
  for_each_n(pol, countAt(0), morton, cv);
}
```
and in `LevelSet`:
```cpp
  while (1) {
    VecDH<int> index(1, 0);
    // avoid handing dynamic function pointers to CUDA
    for_each_n_wrapper<Func>(
        policy, maxMorton + 1,
        ComputeVerts<Func>({vertPos.ptrD(), index.ptrD(), gridVerts.D(), sdf,
                            bounds.min, gridSize + 1, spacing, level}));
```

I guess another specialization is needed to match the type of the closure given by the WASM bindings (a cast function pointer, rather than `std::function`). Unless those happen to be treated the same.

**Comment by geoffder** [comment]
I've run into a strange issue with build artifact linking in my OCaml bindings where Collider refused to be found.
```
/usr/bin/ld: c/libmanifold.a(impl.cpp.o): in function `manifold::Manifold::Impl::Update()':
impl.cpp:(.text+0x2fad): undefined reference to `manifold::Collider::UpdateBoxes(manifold::VecDH<manifold::Box> const&)'
```
This is with bundling the archive in alongside all of the others generated in a build (none of the others gave me this issue). After trying what felt like everything, and making sure that the cmake setup for it was the same as utilities etc, I've given up and moved `collider.h` and `collider.cpp` right into the manifold source directory for now. 

Not so much context I know, but any idea on why collider would be special? I don't *think* I missed anything obvious, but I'm tired of looking at it for now :sweat_smile:. By the way, I've done the build with `add_compile_options(-fPIC)` because I am relocating all of the artifacts into one place to pack in with the binding library. Not sure if I can do anything to avoid it or not, but that got me to being able to build it this far at least.

**Reply by geoffder** [reply]
Reading into it, I wonder if `—whole-archive` would help to just pack in everything to `libmanifoldc.a` and avoid having to move all the archives and link with `-fPIC` as well as remedy this skipped collider situation. Will try and see later. Obviously my lack of non-abstracted away linking experience is biting me now.

**Reply by geoffder** [reply]
After trying it, and reading more, it is infact not the answer to wanting everything packed in to `libmanifoldc.a`. I'd have to use a post build tool like `ar` it seems. 

On the organization of the simple sub-libraries like collider, is there a particular reason why they need to be build as separate shared/static libraries? It seems like making them with the [OBJECT](https://cmake.org/cmake/help/latest/command/add_library.html#object-libraries) (as opposed to STATIC/SHARED) tag, it would make it possible to include them in other libraries directly, rather than packing them into distinct libraries.

**Reply by elalish** [reply]
Honestly, I have less experience with build systems than you do. I guarantee I didn't give this question any thought when I set up this project. If you think it can be improved, I'm all ears.

**Reply by pca006132** [reply]
This is weird, no idea why it will cause such a problem. I can look into it a bit later.

**Reply by geoffder** [reply]
At the moment I'm using `OBJECT` on the little sublibraries, but it seems to require `-fPIC` when CUDA is on (it complains and recommends using the flag), since I guess `OBJECT` works by a sort of relocation that CUDA doesn't like. Seems like there might be less fuss if `collider`, `polygon`, `utilities`, and `graphlite` were moved under the yoke of the manifold library. I'm not sure how much of an impact `-fPIC` has on manifold in particular though.

**Comment by geoffder** [comment]
Still plugging away, but I've implemented *most* of the bindings in OCaml so far (also still have some API decisions to make).

Been trying out replicating some of the examples from the OSCADml repo, though since the mesh is already being generated with OCADml tools, there isn't much for Manifold to do other than verify the meshes and maybe a difference to peak inside (like in this second example).
![image](https://user-images.githubusercontent.com/14910691/203883391-8035b919-811e-4e7c-a858-9627b33d4535.png)
![image](https://user-images.githubusercontent.com/14910691/203884057-3bc5764b-2c49-4ec2-a05a-677fd6892fc7.png)
